docs/proposals/20210811-podmarker.md

```diff
@@ -0,0 +1,199 @@
+- having label "app: nginx"
+- running in the nodes with label "arch: x86"
+- its probe result is "succeeded"
+****
+
+### Specification
+**About Spec.Strategy**
+
+- `Strategy.Replicas`:
+  - `Strategy.Replicas` can ensure that the number(or percentage) of marked Pods `>=` `Replicas`, unless:
+    - `len(satisfiedPods) < Spec.Strategy.Replicas` when `Spec.Strategy.ConflictPolicy` is `"Overwrite"`.
+    - `len(satisfiedPods) - len(conflictPods) < Spec.Strategy.Replicas` when `Spec.Strategy.ConflictPolicy` is `"Ignore"`.
+  - The number(or percentage) of marked Pods `>` `Replicas` iff the number of already marked Pods `>` `Replicas`.
+  - If `Strategy.Replicas` is modified smaller, the PodMarker will shrink the number of marked Pods, and only delete the marks marked by PodMarker;
+  - **Users can remove the marks marked by PodMarker by setting `.Strategy.Replicas = 0`.**
+
+- `Strategy.ConflictPolicy`:
+  - If `ConflictPolicy` is `"Overwrite"`, the labels or annotations, which have the same keys but different values as `spec.MarkItems`,  will be over-written.
+  - If `ConflictPolicy` is `"Ignore"`, the PodMarker will treat these conflicting Pods as unmatched Pods.
+
+**About Spec.Requirements**
+- **A `Pods` is matched if and only if it satisfies all the listed requirements.**
+
+- `Requirements[x].Probe`:
+  - `ProbeRequirements[x].Probe` is a user-defined probe, `PodMarker` will execute this probe and get a result(`succeeded` or `failed`).
+  - if the probe result matches with `ProbeRequirement.Expectation`, `ShouldStatify.ProbeRequirements[x]` will be satisfied.
+  - Users can define more than one probe requirement. `Requirement.ProbeRequirements` is satisfied iff all probe requirements are satisfied.
+
+**About Spec.SortRuleForMatchedPods**
+- `SortRuleForMatchedPods` is an ordered list, which describe the priority of matched Pods when marking them;
+- `SortRuleForMatchedPods[i]` has much higher priority than `SortRuleForMatchedPods[j]` if `i < j`;
+- If `len(SortRuleForMatchedPods) > 1`, this sorting rule is similar to [lexicographic order](https://en.wikipedia.org/wiki/Lexicographic_order).
+****
+
+### Implementation Plan
+#### WebHook Validation
+1. Validate `.Metadata`
+2. Validate `.Spec.Strategy`:
+- check whether `Replicas` is a number or percentage;
+- check two Policies.
+3. Validate `.Spec.Requirements`
+- check `PodSelector` and `NodeSelector;
+- check `ProbeRequirements[x].Probe`;
+- check container names;
+- check whether `.ProbeRequirements[x].Expectation` is succeeded or failed.
+4. Validate `.Spec.SortRuleForMatchedPods`
+5. Validate `.Spec.PodMarkerMarkItems`
+
+#### Controller
+**Recognize the Pods Marked by PodMarker**
+
+PodMarker will mark an extra annotation `kruise.io/marked-by-podmarker=<podmarker-name>` on Pods.
+
+**Implement in Kruise-Daemon**
+
+We are going to implement the controller logic in `kruise-daemon` due to the following problems:
```

Comment:
Except probing, i don't think it is necessary to depend on kruise-daemon,  a controller in kruise-manager is enough

Reply:
I agree